### ISSUE TITLE
fontconfig ignoring fonts installed with nix-env

### ISSUE THREAD

**Issue by kirelagin** [NixOS/nixpkgs#8318]
I’m not really sure why but, well, fontconfig ignores my `~/.nix-profile/share/fonts`.

~~~~Console
› /nix/store/bc3sk492zcv744qfj8nnrlbw9yl6gmnk-fontconfig-2.11.1/bin/fc-cache -v | grep nix-profile
›
~~~~

even though its config tells it to look there:

~~~~Console
› cat /nix/store/bc3sk492zcv744qfj8nnrlbw9yl6gmnk-fontconfig-2.11.1/etc/fonts/fonts.conf | grep nix-profile                                                                               <1>
<dir>~/.nix-profile/lib/X11/fonts</dir><dir>~/.nix-profile/share/fonts</dir><dir>/nix/var/nix/profiles/default/lib/X11/fonts</dir><dir>/nix/var/nix/profiles/default/share/fonts</dir></fontconfig>
~~~~

Any ideas?

**Comment by kirelagin** [comment]
By the way, my system fontconfig (which includes `~/.config/fontconfig/fonts.conf` and is told there to check the `~/.nix-profile/share/fonts` directory) works fine:

~~~~Console
› fc-cache -v | grep nix-profile
/home/kirrun/.nix-profile/share/fonts: skipping, existing cache is valid: 0 fonts, 1 dirs
/home/kirrun/.nix-profile/share/fonts/opentype: skipping, existing cache is valid: 7 fonts, 0 dirs
Re-scanning /home/kirrun/.nix-profile/share/fonts: caching, new cache contents: 0 fonts, 1 dirs
~~~~

**Comment by lethalman** [comment]
@kirelagin I haven't understood the difference between the first comment and the second.

**Comment by kirelagin** [comment]
@lethalman The first one uses `fc-cache` from the nix store, the second one uses the system `fc-cache` (this is not NixOS).

**Comment by kirelagin** [comment]
Ok, I got it.
The problem is fontconfig being horribly inconsistent. It reads my _system_ `fonts.conf` but assumes that the config directory is in `$out`, so when it sees `<include ignore_missing="yes">conf.d</include>` in my _system_ config it goes to `$out/etc/fonts/conf.d` and reads configs from there.

Since here on Gentoo user config is included from `/etc/fonts/conf.d/50-user.conf`, my user config is actually never getting included.

It should definitely be reading files only from one source. As far as I can tell, by default it would be reading everything (including `fonts.conf`) from `$out/etc/fonts/` and this does make sense in my opinion, but [this patch](https://github.com/NixOS/nixpkgs/blob/4eeb4c88aab202f6841d6e2d959f0b6ece67c908/pkgs/development/libraries/fontconfig/config-compat.patch) breaks this. @vcunat

**Comment by kirelagin** [comment]
If the intention was to let NixOS have its completely separate fonts configuration, then, keeping in mind the fact that we definitely want working fontconfig on non-NixOS, it might be better to:

* drop this patch all together, letting fontconfig get all its configuration from `$out` by default,
* generate custom config for NixOS (probably somewhere in the store; why is it being add to `/etc` via `config.environment` right now?),
* make fontconfig read it by means of `$FONTCONFIG_FILE` and `$FONTCONFIG_PATH`.

**Comment by edolstra** [comment]
We don't want NixOS to have separate fonts configuration. The only reason for those patches is to work around fontconfig version incompatibilities (requiring us to have different config files for 2.10 and 2.11, IIRC).

I guess we just need to build Fontconfig with `--sysconfdir=/etc` to ensure `<include ignore_missing="yes">conf.d</include>` is resolved properly on Gentoo.

**Comment by lethalman** [comment]
@edolstra yes that should be sufficient. We use absolute paths in our fonts.conf so it won't be a problem.

**Comment by kirelagin** [comment]
This probably means that fontconfig in nixpkgs shouldn’t ship any kind of `fonts.conf`.
On other distros the config will be already present in `/etc` and on NixOS the fontconfig module will generate one.

**Comment by jagajaga** [comment]
Ping all.

**Comment by kirelagin** [comment]
So, what this all fontconfig 2.10 thing is about? I don’t see it being used anywhere in nixpkgs.
Can’t we just drop this backward compatibility, the patch and forget about the issue?
@edolstra

**Comment by vcunat** [comment]
That should be unrelated to the problems described in the title. It's to support running stuff *built* previously against fontconfig 2.10.

**Comment by SevereOverfl0w** [comment]
Just as a tide-over for others experiencing this, I modified my /etc/fonts/fonts.conf, to use the absolute path to the conf.d folder, it isn't a great fix, but it saves me from experiencing Oswalds Font everywhere!

**Comment by johan-boule** [comment]
SevereOverfl0w : Thank you for the workaround, it worked for me too.

**Comment by knedlsepp** [comment]
I'm on a non-NixOS system trying to view pdf's with some pdfs that stem from Microsoft office, I could make it work via
```
nix-env -iA nixpkgs.corefonts
```
 and symlinking the fonts to `~/.local/share/fonts` via:
 ```
ln -sf ~/.nix-profile/share/fonts/ ~/.local/share/fonts/nix-fonts
```

**Comment by doronbehar** [comment]
It's too bad this issue is still open, I'll confirm though that @knedlsepp's workaround is working so thanks!

**Comment by knedlsepp** [comment]
I guess https://github.com/rycee/home-manager might also be a viable Option.
